No-coll Movement

spectre1989 opened the pull request:
## Summary
No-collision movement on the server now matches the client.

### Additions/modifications proposed in this pull request:
The major change that made this work was how the server handles inputs from the client.
- Using the send_id in the input packet, the server is now able to figure out if any of the control state changes in the packet have been seen before.
- Inputs are then queued in the entity input state, and each tick they are processed. The server doesn't need to do any calculations about time, everything it needs is in the control state changes.

### Things I'm Unsure of:
- The way I've translated some of the RE movement code to server-side, it seems to work but I'm not sure how I'd fully test all code paths
- Compiler warnings, which can be ignored and which shouldn't
- When I tested with 2 clients with input and movement logging enabled, the sheer quantity of logs seemed to make my server fall over, if this is unacceptable then perhaps need a way to select which specific entity to log input/movement for?

spectre1989 committed:
wip movement stuff

spectre1989 committed:
bunch of logging for control state unpacking

spectre1989 committed:
more wip input processing, figuring out how to bundle control state changes

spectre1989 committed:
control state changes for keys tested and appear to be working, key press time appears to be working, now to try calculating input velocity and position delta

spectre1989 committed:
local input velocity in x and z appears to be correct

spectre1989 committed:
untested attempt at wasd nocoll movement

spectre1989 committed:
added pitch/yaw to control state changes, movement does not yet appear to match up

spectre1989 committed:
trying to get key press time correct

spectre1989 committed:
more movement logging, and input handling closer to client

spectre1989 committed:
wasd movement seems to work in nocoll! Does desync eventually but suspect this is rounding errors and/or quantisation. With 2 players server falls over quickly, is it the obscene amount of logging??

spectre1989 committed:
commented out a bunch of logging that was making the server chug with > 1 player, need to make a slash command to turn it on for a specific player for debugging purposes

spectre1989 committed:
part way through refactor...

spectre1989 committed:
fixing more compile errors post-refactor

spectre1989 committed:
post-refactor it now compiles, plenty of testing, todos, and tidying up to be done now

spectre1989 committed:
a couple of issues remain I think caused by the up/down controls, but otherwise working well

spectre1989 committed:
cleaned up a bunch of todos

spectre1989 committed:
couple of tiny things

spectre1989 committed:
refactor csc processing loop so it will skip any old ones rather than iterating
moved input logging to csc loop

spectre1989 committed:
tidied up movement logging, should now be close to what /controldebug 1 on client generates

spectre1989 committed:
refactor complete, needs testing, and probably looking at warnings

spectre1989 committed:
fixed final input velocity always 0 caused by stupid typo
fixed logging of extended pitch/yaw
fixed up/down in final pass of RE source

spectre1989 committed:
Merge branch 'develop' into movement

# Conflicts:
#	Projects/CoX/Common/GameData/Movement.cpp
#	Projects/CoX/Common/Messages/Map/InputState.cpp

spectre1989 committed:
tidying up

spectre1989 committed:
added a todo about input acks

broxen commented:
Regarding `When I tested with 2 clients with input and movement logging enabled, the sheer quantity of logs seemed to make my server fall over, if this is unacceptable then perhaps need a way to select which specific entity to log input/movement for?`

Perhaps a per-character variable `m_log_movement` to check before sending logs out. Not sure that's the cleanest way, but it would work.

spectre1989 commented:
> Regarding `When I tested with 2 clients with input and movement logging enabled, the sheer quantity of logs seemed to make my server fall over, if this is unacceptable then perhaps need a way to select which specific entity to log input/movement for?`
> 
> Perhaps a per-character variable `m_log_movement` to check before sending logs out. Not sure that's the cleanest way, but it would work.

And enable it per-character with a /command?

broxen commented:
> And enable it per-character with a /command?

Precisely `/logmovement {optional: $target}` or something. Variable may be best in the Entity class, as different types of entities may have movement (Monorail, Blimp, Pedestrians, etc)

broxen commented on Projects/CoX/Common/GameData/StateStorage.h:
```diff
@@ -40,114 +104,92 @@ class TimeState
+    bool        m_has_keys = false;
+    bool        m_keys[6] = {};
+
+    // whether this state change contains the client pitch/yaw which are sent
```
These may actually be camera pitch/yaw, and perhaps the angle deviates due to some kind of rounding error with mouse movement? Worth testing to see.

broxen commented on Projects/CoX/Common/GameData/StateStorage.h:
```diff
@@ -40,114 +104,92 @@ class TimeState
-        return &m_inp_states.back()-1;
-    }
+    // sent by client, not sure what it's for
+    float                           m_velocity_scale = 1.0f;
```
I believe this should increase the speed of all movement if the client has changed it's value.

spectre1989 committed:
Merge branch 'develop' into movement

broxen commented on Projects/CoX/Common/GameData/Movement.cpp:
```diff
@@ -49,195 +58,571 @@ SurfaceParams g_world_surf_params[2] = {
 
-        if(vel.z < 0.0f)
-            vel_scale_copy *= motion->m_backup_spd;
+static void playerMotionSetInputVelocity(Entity* player, const TickState* tick_state) // based on pmotionSetVel()
```
~~The var name `player` is misleading here. A player is always represented by an Entity, but an Entity may or may not be a Player.~~

Disregard my comment. This may not matter if we never use this method for other entity types (Critters, Blimps, Cars, etc). Input was the key word here and I missed it =P

spectre1989 replied:
Yeah I'm making an assumption that this code only runs for players, if that's not the case it'll need changing

spectre1989 committed:
slash commands for toggling input/movement logging

spectre1989 committed:
Merge branch 'develop' into movement

spectre1989 commented:
> > And enable it per-character with a /command?
> 
> Precisely `/logmovement {optional: $target}` or something. Variable may be best in the Entity class, as different types of entities may have movement (Monorail, Blimp, Pedestrians, etc)

ok, I've added it, but can't test it yet because of the db issues we're having at present

broxen commented:
Give this is another go pulling the latest from develop head.

broxen commented on Projects/CoX/Servers/MapServer/SlashCommand.cpp:
```diff
@@ -257,6 +259,8 @@ static const SlashCommand g_defined_slash_commands[] = {
     {{"clientConsoleOutput"}, "Send message to ingame (~) console", cmdHandler_SendConsolePrint, 9},
     {{"clearTarget"}, "Clear current target", cmdHandler_ClearTarget, 9},
     {{"startTimer"}, "Create a small timer", cmdHandler_StartTimer, 9},
+    {{"toggleInputLog"}, "Enable input logging for a player", cmdHandler_ToggleInputLog, 9},
```
Smart idea.

broxen commented on Projects/CoX/Servers/MapServer/SlashCommand.cpp:
```diff
@@ -1306,6 +1310,36 @@ void cmdHandler_StartTimer(const QStringList &params, MapClientSession &sess)
 
+void cmdHandler_ToggleInputLog(const QStringList &params, MapClientSession &sess)
+{
+    Entity* target = getTargetEntity(sess);
```
Following the format of `/teleport` may allow admins to type in a name of a player that they can't target directly (eventually, using UserRouter, another zone etc).

It makes several checks for name and target to avoid duplicate name issues where possible.

```cpp
    QString name = sess.m_ent->name();
    if(!params.isEmpty())
        name = params.join(" ");

    // getTargetByNameOrIdx will always return a valid Entity, or self
    Entity *tgt = getCmdTargetByNameOrIdx(sess, name);
    name = tgt->name(); // make sure we have the final name
```

spectre1989 replied:
I've updated it to use that with the latest commit, and tested with 2 clients, switching logging between myself and another player, seems to work.

spectre1989 committed:
Merge branch 'develop' into movement

spectre1989 commented:
> Give this is another go pulling the latest from develop head.

Yep that worked, I'll finish up the logging stuff tomorrow

spectre1989 committed:
input/movement logs only created for explicit entities to avoid swamping the server
/toggle[Input/Movement]Log gets target by name rather than selected target (or self if there is none)

broxen commented on Projects/CoX/Common/GameData/StateStorage.h:
```diff
@@ -26,6 +26,70 @@ enum BinaryControl
+    {
+      uint8_t key_state;            // control_id 0-5
+      float angle;                  // control_id 6-7
+      struct
```
It probably doesn't really matter, but `-wPedantic` is throwing roughly a dozen warnings about Anonymous structs here. Let's name this.

`warning: ISO C++ prohibits anonymous structs [-Wpedantic]`

spectre1989 replied:
Ah yeah, I meant to do that, will do

broxen commented on Projects/CoX/Common/GameData/Movement.cpp:
```diff
@@ -49,195 +58,601 @@ SurfaceParams g_world_surf_params[2] = {
 };
 
-void roundVelocityToZero(glm::vec3 *vel)
+static bool floatEquals(float a, float b)
```
We don't actually use this anywhere. Is there a future use planned, or should we just remove it?

spectre1989 replied:
Oops, just left over from when I was using it, and then I wasn't...

broxen commented on Projects/CoX/Common/GameData/Movement.cpp:
```diff
@@ -49,195 +58,601 @@ SurfaceParams g_world_surf_params[2] = {
+                                qCDebug(logInput, "doing tick, controls_disabled=%d, time_diff_1=%u, time_diff_2=%u, velocity_scale=%hhu", csc.controls_disabled, csc.time_diff_1, csc.time_diff_2, csc.velocity_scale);
+                            }
+
+                            // do a tick!
```
Errant comment? or TODO?

spectre1989 replied:
It's a comment - control id 8 is "end of tick" which means a tick can be executed. Should I change this comment to something else?

broxen replied:
I would make this more verbose so it's clear that the behavior is already there, and this isn't a TODO for future code or something.

broxen commented on Projects/CoX/Common/GameData/StateStorage.cpp:
```diff
@@ -60,34 +60,28 @@ void TimeState::dump()
+        for (bool key : m_keys)
+        {
+            if (key)
+            {
```
For code consistency, let's omit brackets on single line code blocks, except for where nested. Referencing our coding guidelines:
```cpp
for (bool key : m_keys)
{
    if (key)
        return true;
}

spectre1989 replied:
MOTHER OF GOD

broxen reviewed (changes requested):
Thank you for working on this spectre!

I'm running into a few issues after building this PR
- server position doesn't actually appear to be updated, I'm not sure if I need to toggle server authority, but `/movement` didn't seem to have any effect.
- using powers sometimes disables movement altogether. In particular "Sprint"
- turning with the mouse or keyboard `q` or `e` has significant latency. It takes the client several seconds to update the current player's rotation.
- Other clients do see player rotation, but with the same delay as seen locally.
- Other clients don't see movement at all.

spectre1989 committed:
removed unused function floatEquals()
fixed warning about anonymous struct in control state change struct
removed some braces for single line ifs

spectre1989 commented:
> Thank you for working on this spectre!
> 
> I'm running into a few issues after building this PR
> 
> * server position doesn't actually appear to be updated, I'm not sure if I need to toggle server authority, but `/movement` didn't seem to have any effect.
> * using powers sometimes disables movement altogether. In particular "Sprint"
> * turning with the mouse or keyboard `q` or `e` has significant latency. It takes the client several seconds to update the current player's rotation.
> * Other clients do see player rotation, but with the same delay as seen locally.
> * Other clients don't see movement at all.

I can't repro those issues locally, are you running the server on your local machine or a remote one? I've only tried running the server on the same machine as the client.

Server authority is likely still broken, I couldn't get it to work before and I've not looked at fixing it yet. I was thinking perhaps (once we're happy with it) we should turn on server authority by default for no-coll movement only, but obviously as all other movement is unimplemented leave it off for that.

spectre1989 commented:
I'm still working on an update for this - hope it's alright with you @broxen, I'm integrating stuff from your movement branch as you've got the whole structure and flow of entMotion/entWalk/etc.

spectre1989 committed:
integrated broxens movement work

broxen commented:
Absolutely, feel free to use that code. Tread carefully though, I can't remember what I did there =P

spectre1989 committed:
spoofing movement

spectre1989 commented:
Ok as discussed I've integrated a lot of the code from your movement branch, but I've left out entWorldCollide. For now I'm spoofing player movement so that it roughly matches player movement in the starting area that I was testing with. I've also renamed a lot of the functions to be consistent with yours, I figure that'll make it less confusing to work with going forward if I stop renaming things!

broxen commented:
CI is failing due to the use of compare in the body parts changes Nem made recently. I'll try to push a fix for that tonight.

```cpp
/home/travis/build/Segs/Segs/Projects/CoX/Common/GameData/bodypart_definitions.cpp: In member function ‘BodyPart_Data* BodyPartsStorage::getBodyPartFromName(const QByteArray&)’:
/home/travis/build/Segs/Segs/Projects/CoX/Common/GameData/bodypart_definitions.cpp:120:27: error: ‘class QByteArray’ has no member named ‘compare’
         if ( 0==bp.m_Name.compare(name,Qt::CaseInsensitive) )
                           ^~~~~~~
```

nemerle commented:
That problem with CI is likely caused by the older Qt version on the server.

cajoebestrotheray commented:
> CI is failing due to the use of compare in the body parts changes Nem made recently. I'll try to push a fix for that tonight.
> 
> ```c++
> /home/travis/build/Segs/Segs/Projects/CoX/Common/GameData/bodypart_definitions.cpp: In member function ‘BodyPart_Data* BodyPartsStorage::getBodyPartFromName(const QByteArray&)’:
> /home/travis/build/Segs/Segs/Projects/CoX/Common/GameData/bodypart_definitions.cpp:120:27: error: ‘class QByteArray’ has no member named ‘compare’
>          if ( 0==bp.m_Name.compare(name,Qt::CaseInsensitive) )
>                            ^~~~~~~
> ```

Is there anything I need to do for this? Like merge upstream and push to my branch or something?

broxen commented:
We might be able to adjust CI to to use a newer QT version. Also Nem mentioned in discord a possible cmake change that would look for the latest version.

I've been pretty slammed with work and haven't had a chance to look at it, but those are two possible solutions

spectre1989 committed:
Merge branch 'develop' into movement

spectre1989 committed:
making comment less vague/confusing

broxen commented:
This is an enormous step in the right direction. Fantastic work spectre!
